File: routers/github.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional, Any
import httpx

router = APIRouter(prefix="/github", tags=["GitHub"])

class GitHubScanRequest(BaseModel):
    owner: str
    repo: str
    token: Optional[str] = None
    path: str = ""

class GitHubContentRequest(BaseModel):
    owner: str
    repo: str
    token: Optional[str] = None
    path: str

class FileInfo(BaseModel):
    name: str
    path: str
    type: str
    download_url: Optional[str] = None

@router.post("/scan", response_model=List[FileInfo])
async def scan_github(data: GitHubScanRequest):
    """
    Scans a GitHub repository for files.
    """
    url = f"https://api.github.com/repos/{data.owner}/{data.repo}/contents/{data.path}"
    headers = {"Accept": "application/vnd.github.v3+json"}
    if data.token:
        headers["Authorization"] = f"token {data.token}"
        
    async with httpx.AsyncClient() as client:
        try:
            response = await client.get(url, headers=headers)
            if response.status_code != 200:
                raise HTTPException(status_code=response.status_code, detail=response.json().get("message", "GitHub API Error"))
            
            contents = response.json()
            if not isinstance(contents, list):
                # It's a single file
                contents = [contents]
                
            return [FileInfo(
                name=item["name"],
                path=item["path"],
                type=item["type"],
                download_url=item.get("download_url")
            ) for item in contents if item["type"] == "file" and item["name"].endswith(".md")]
            
        except httpx.HTTPError as e:
            raise HTTPException(status_code=500, detail=str(e))

@router.post("/content")
async def get_github_content(data: GitHubContentRequest):
    """
    Fetches the content of a file from GitHub.
    """
    url = f"https://api.github.com/repos/{data.owner}/{data.repo}/contents/{data.path}"
    headers = {"Accept": "application/vnd.github.v3.raw"}
    if data.token:
        headers["Authorization"] = f"token {data.token}"
        
    async with httpx.AsyncClient() as client:
        try:
            response = await client.get(url, headers=headers)
            if response.status_code != 200:
                raise HTTPException(status_code=response.status_code, detail="Failed to fetch file content")
            
            return {"content": response.text}
            
        except httpx.HTTPError as e:
            raise HTTPException(status_code=500, detail=str(e))

File: src/routers/github.ts
import { Router, Request, Response } from 'express';
import axios from 'axios';
import { GitHubScanRequestSchema, GitHubContentRequestSchema, FileInfoSchema } from '../types/schemas';

const router = Router();

/**
 * Scans a GitHub repository for files.
 */
router.post('/scan', async (req: Request, res: Response) => {
  const validation = GitHubScanRequestSchema.safeParse(req.body);
  if (!validation.success) {
    return res.status(422).json({ detail: validation.error.errors });
  }

  const { owner, repo, token, path } = validation.data;
  const url = `https://api.github.com/repos/${owner}/${repo}/contents/${path}`;
  const headers: Record<string, string> = {
    'Accept': 'application/vnd.github.v3+json',
    'User-Agent': 'Delema-API-NodeJS'
  };
  if (token) {
    headers['Authorization'] = `token ${token}`;
  }

  try {
    const response = await axios.get(url, { headers });
    let contents = response.data;
    if (!Array.isArray(contents)) {
      contents = [contents];
    }

    const files = contents
      .filter((item: any) => item.type === 'file' && item.name.endsWith('.md'))
      .map((item: any) => ({
        name: item.name,
        path: item.path,
        type: item.type,
        download_url: item.download_url
      }));

    return res.json(files);
  } catch (error: any) {
    const status = error.response?.status || 500;
    const detail = error.response?.data?.message || error.message;
    return res.status(status).json({ detail });
  }
});

/**
 * Fetches the content of a file from GitHub.
 */
router.post('/content', async (req: Request, res: Response) => {
  const validation = GitHubContentRequestSchema.safeParse(req.body);
  if (!validation.success) {
    return res.status(422).json({ detail: validation.error.errors });
  }

  const { owner, repo, token, path } = validation.data;
  const url = `https://api.github.com/repos/${owner}/${repo}/contents/${path}`;
  const headers: Record<string, string> = {
    'Accept': 'application/vnd.github.v3.raw',
    'User-Agent': 'Delema-API-NodeJS'
  };
  if (token) {
    headers['Authorization'] = `token ${token}`;
  }

  try {
    const response = await axios.get(url, { headers, responseType: 'text' });
    return res.json({ content: response.data });
  } catch (error: any) {
    const status = error.response?.status || 500;
    const detail = error.response?.data?.message || 'Failed to fetch file content';
    return res.status(status).json({ detail });
  }
});

export default router;
